jms/src/main/scala/akka/stream/alpakka/jms/JmsConsumerStage.scala

```diff
@@ -162,32 +162,29 @@ final class JmsTxSourceStage(settings: JmsConsumerSettings)
       protected def createSession(connection: Connection, createDestination: Session => javax.jms.Destination) = {
         val session =
           connection.createSession(true, settings.acknowledgeMode.getOrElse(AcknowledgeMode.SessionTransacted).mode)
-        new JmsTxSession(connection, session, createDestination(session), settings.destination.get)
+        new JmsConsumerSession(connection, session, createDestination(session), settings.destination.get)
       }
 
       protected def pushMessage(msg: TxEnvelope): Unit = push(out, msg)
 
       override protected def onSessionOpened(jmsSession: JmsConsumerSession): Unit =
         jmsSession match {
-          case session: JmsTxSession =>
+          case session: JmsSession =>
             session.createConsumer(settings.selector).onComplete {
               case Success(consumer) =>
                 consumer.setMessageListener(new MessageListener {
 
-                  var listenerStopped = false
-
                   def onMessage(message: Message): Unit =
-                    if (!listenerStopped)
-                      try {
-                        val envelope = TxEnvelope(message, session)
-                        handleMessage.invoke(envelope)
-                        val action = session.commitQueue.take()
-                        action(envelope)
-                      } catch {
-                        case _: StopMessageListenerException => listenerStopped = true // Tombstone.
-                        case e: IllegalArgumentException => handleError.invoke(e) // Invalid envelope. Fail the stage.
-                        case e: JMSException => handleError.invoke(e)
-                      }
+                    try {
+                      val envelope = TxEnvelope(message, session)
+                      handleMessage.invoke(envelope)
+                      val action = Await.result(envelope.commitFuture, settings.ackTimeout)
```

Comment:
Whose thread is blocked here? How many outstanding commits can it take?

Reply:
This is the JMS session thread (which is blocking anyway... previously just on the blocking queue). The number of outstanding commits is the number of sessions we open.

In our production we do see anomalies and deviations from the JMS spec in edge cases when using the previous implementation of blocking queue. Switching to futures makes it more lenient to such anomalies and has 0 incidents over the past month. We could use either this or the Java `CompletableFuture`. Both cases we do block in similar fashion, though.